code/design_pattern/decorator_pattern/decorator_pattern.cpp

```diff
@@ -0,0 +1,101 @@
+#include <iostream>
+#include <string>
+
+class IBeverage
+{
+public :
+    IBeverage() 
+    {
+        description = "Unkown beverage";
+    }
+    virtual std::string getDescription()
+    {
+        return description;
+    }
+    virtual double cost() = 0;
+protected :
+    std::string description;
+};
+
+class HouseBlend : public IBeverage
+{
+public :
+    HouseBlend()
+    {
+        description = "House Blend Coffee";
+    }
+    
+    double cost()
+    {
+        return 0.89;
+    }
+};
+
+class CondimentDecorator : public IBeverage
+{
+public :
+    CondimentDecorator(IBeverage* bw)
+    {
+        this->beverageWrapper = bw;
+    }
+    
+    std::string getDescription()
+    {
+        return this->beverageWrapper->getDescription();
+    }
+
+    double cost()
+    {
+        return this->beverageWrapper->cost();    
+    }
+    
+    ~CondimentDecorator()
+    {
+        delete beverageWrapper;
+    }
+private :
+    IBeverage* beverageWrapper;
+};
+
+class Mocha : public CondimentDecorator
+{
+public :
+    Mocha(IBeverage* beverage) : CondimentDecorator(beverage) {}
+
+    std::string getDescription()
+    {
+        return CondimentDecorator::getDescription() + ", Mocha ";
+    }
+
+    double cost()
+    {
+        return 0.20 + CondimentDecorator::cost();
+    }
+};
+
+class Whip : public CondimentDecorator
+{
+public :
+    Whip(IBeverage* beverage) : CondimentDecorator(beverage) {}
+
+    std::string getDescription()
+    {
+        return CondimentDecorator::getDescription() + ", Whip ";
+    }
+
+    double cost()
+    {
+        return 0.10 + CondimentDecorator::cost();
+    }
+};
+
+int main() 
+{
+    IBeverage* beverage = new HouseBlend();
+    std::cout << beverage->getDescription() + " $" << beverage->cost() << std::endl;
+    beverage = new Mocha(beverage);
```

Comment:
You have to delete the beverage before you set it to a new one in either case. Here I would seriously recommend using unique_ptr.

```C++
std::unique_ptr<IBeverage> beverage(new HouseBlend());
std::cout << beverage->getDescription() + " $" << beverage->cost() << std::endl;
beverage.reset(new Mocha(beverage.get()));
std::cout << beverage->getDescription() + " $" << beverage->cost() << std::endl;
beverage.reset(new Whip(beverage.get()));
std::cout << beverage->getDescription() + " $" << beverage->cost() << std::endl;
```

Reply:
Remember, in C++ there is no automatic garbage collection.

Reply:
Thank you, I really appreciate the feedback. :D

Reply:
I can't seem to make unique_ptr work, so I resorted to making a tmp pointer to delete after wrapping beverage with mocha, but deleting said pointer causes a crash

Reply:
What are the errors you are getting for the unique_ptr? Are you using C++11? Are you including `<memory>`? If you are making a temp pointer, then do not delete anything, until the very end. Delete only once all pointers that point to the same thing, and delete once for each pointer otherwise.

Reply:
Honestly I've been looking at the code for a while, I can't seem to figure out where the memory leak is. If I just delete beverage in the end of the main function, shouldn't that cover it?